### PR TITLE
AES perf improvements

### PR DESCRIPTION
**Pull request by daverodgman**
## Description

This improves block-XOR performance on architectures with 64-bit registers, which benefits mostly AES (presumably because use of AESCE / AESNI is fast so the XOR part is more significant than other ciphers).

I've modified the AES-CBC implementation to make better use of this and also eliminated some unnecessary memcpy's, which has improved AES-CBC performance everywhere (around 7%-11%).

Performance uplift is quite varied depending on compiler (clang vs gcc), optimisation level (O2 vs Os) and mode (all measurements on M1 Pro). Everything is same-or-faster (only clang -Os XTS is the same; everything else sees some uplift). gcc gets most benefit (around 30% average over the different modes and optimisation levels). clang averages around 8% uplift overall. Biggest win is AES-XTS (61-85% on gcc -O2); GCM gets 11-58% uplift (gcc -Os is 44-58%; everything else is 11-17%).

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** provided
- [x] **backport** not required - not a bugfix
- [x] **tests** covered by existing tests

### CODE REVIEW

**Review by tom-daubney-arm — DISMISSED** [review]
This is a net win which I think is worth the small regression in AES-CBC-256.

LGTM.

**Comment by daverodgman** [comment]
Turns out the CBC regression is because, unlike XTS, it does an xor of the form `xor(dest, src, dest)`, whereas XTS does `xor(dest, src1, src2)`. Using a temp variable can improve this.

In further experiments, there's also (a lot of) further uplift available from NEON.

**Comment by daverodgman** [comment]
Had a quick experiment with reversing the direction of decryption, which allows it to avoid the temporary `memcpy()` so looks quicker on paper, but this turns out to be slower - I guess the hardware is slightly less efficient traversing memory backwards.

**Comment by gilles-peskine-arm** [comment]
> I guess the hardware is slightly less efficient traversing memory backwards

What hardware did you try? That looks like an effect of prefetching. On Arm, even as far back as [Cortex-A9](https://developer.arm.com/documentation/ddi0388/i/level-1-memory-system/data-prefetching/data-prefetching), the prefetcher detects backward traversal. On Intel, IIRC, prefetching came earlier and was forward-only but started to handle both directions roughly around the same time as the Cortex-A9 came out.

**Comment by daverodgman** [comment]
Yes, prefetcher is the obvious candidate, but this was on M1 Pro which has very strong memory performance and presumably a very capable prefetcher, so I don't have a good idea what the explanation would be.

**Comment by daverodgman** [comment]
Looking more closely at decryption perf, I still can't see why changing direction doesn't help.

I did notice that the benchmark program does not actually include decryption in the resulting number (I think I missed this previously and assumed it measured round-trip perf). Measuring decryption separately, I don't think the optimisation here actually helps decryption at all, so I'll remove that.

However, I did look more at why encryption is so much slower (around 3x). It seems to be due to the data dependency from block to block that is only present for encryption; this can't be avoided, but the latest commit does reduce its impact a little bit (this commit improves encryption perf by an additional 0.6 - 2% depending on compiler/flags, if AESCE is enabled, and no significant impact if it isn't).

**Comment by daverodgman** [comment]
Sorry for the churn @tom-daubney-arm - I think this is ready for review - I don't plan any further updates

**Comment by tom-daubney-arm** [comment]
> Sorry for the churn @tom-daubney-arm - I think this is ready for review - I don't plan any further updates

Cool thank you, I will TAL ASAP.

File: library/aes.c [hunk 1]
```diff
@@ -1086,13 +1086,13 @@ int mbedtls_aes_crypt_cbc(mbedtls_aes_context *ctx,
...
-            mbedtls_xor(output, input, iv, 16);
+            mbedtls_xor(temp, input, iv, 16);
 
-            ret = mbedtls_aes_crypt_ecb(ctx, mode, output, output);
+            ret = mbedtls_aes_crypt_ecb(ctx, mode, temp, iv);
+            memcpy(output, iv, 16);
             if (ret != 0) {
                 goto exit;
             }
-            memcpy(iv, output, 16);
```
Comment (tom-cosgrove-arm) [inline library/aes.c:1095]: what's the benefit of moving the memcpy? We're unlikely to hit an error, so don't stall to check `ret` given it's likely to be `0`?

Comment (daverodgman) [reply library/aes.c:1095]: I don't know why, but saw a difference in benchmarks.

File: library/aes.c [hunk 2]
```diff
@@ -1089,19 +1087,18 @@ int mbedtls_aes_crypt_cbc(mbedtls_aes_context *ctx,
...
             if (ret != 0) {
                 goto exit;
             }
-            ivp = output;
 
             input  += 16;
             output += 16;
             length -= 16;
         }
-        memcpy(iv, ivp, 16);
```
[the remark this answers is not in the document]
Comment (daverodgman) [reply library/aes.c]: That said, benchmarking now I see a small benefit for moving the memcpy out of the loop (at least for clang -O0)

[the remark this answers is not in the document]
Comment (daverodgman) [reply library/aes.c]: I'm inclined to favour "-Os" as this is our most important use-case, and restore this change.

**Comment by daverodgman** [comment]
The last commit isn't the neatest, but it adds ~3/1/7% to CBC 128/192/256 with clang -Os on aarch64.

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM - nice, thanks

**Review by tom-daubney-arm — APPROVED** [review]
LGTM

**Pull request merged**